lib/smf/smf.c

```diff
@@ -203,7 +329,27 @@ void smf_set_initial(struct smf_ctx *ctx, const struct smf_state *init_state)
 	ctx->current = init_state;
 	ctx->previous = NULL;
 	ctx->terminate_val = 0;
+#ifdef CONFIG_SMF_UML_HSM
```

Comment:
For `smf_set_initial` you're handling the UML and non-UML paths inline.  But for `smf_set_state`, you've overloaded the function definition.

I think this is easier to read if you organize as an inline choice, but with helper functions.
```
void smf_set_initial()
#ifdef CONFIG_SMF_UML_HSM
    smf_uml_set_initial();
#else
    smf_non_uml_set_initial();
#endif
```
This will also allow you to group the private UML and non-UML functions into a single #ifdef block.

I'm open to suggestions for the naming of the two types of state machines.  `non_uml` isn't great.

Reply:
Good point. I've notices the existing functions use `__unused` instead of `#if` blocks to ignore unused functions, and I was concerned that leaves the UML functions opened to being called in by the original-style code, which could cause problems.

How about UML and classic ;)